NB-11852: Set ReleaseChannel In JumpCloud Password Manager command

gbitarjc opened the pull request:
## Issues
* [NB-11852](https://jumpcloud.atlassian.net/browse/NB-11852) - Desktop App Dogfood

## What does this solve?
Setting release channel for password manager application in user's data.
Supported Operating systems: Linux, Windows and MacOS.
Supported channels: public, beta and dogfood.

## Is there anything particularly tricky?
Yes, the testing

## How should this be tested?
**For Windows**
- Create the below `.ps1` script on your machine
```powershell
# Set $releaseChannel to beta OR dogfood OR public depending on your desired release channel
$releaseChannel = "dogfood"
#------- Do not modify below this line ------

$allowed_values = @("beta", "dogfood", "public")

if (-not ($allowed_values -ccontains $releaseChannel)) {
    Write-Host "Error: Variable `$releaseChannel must be either 'beta', 'dogfood', or 'public'."
    exit 1
}

# Get the current user's SID (Security Identifier)
$loggedUser = Get-WmiObject -Class Win32_ComputerSystem | Select-Object -ExpandProperty UserName
$loggedUser = $loggedUser -replace '.*\\'

# Construct the Registry path using the user's SID
$userSID = (New-Object System.Security.Principal.NTAccount($loggedUser)).Translate([System.Security.Principal.SecurityIdentifier]).Value
$registryPath = "HKLM:\SOFTWARE\Microsoft\Windows NT\CurrentVersion\ProfileList\$userSID"
$loggedOnUserProfileImagePath = Get-ItemPropertyValue -Path $registryPath -Name 'ProfileImagePath'
$filePath = "$loggedOnUserProfileImagePath\AppData\Roaming\JumpCloud Password Manager\data\daemon\releaseChannel.txt"

$directory = Split-Path $filePath
if (-not (Test-Path $directory)) {
    New-Item -ItemType Directory -Path $directory -Force
}

Set-Content -Path $filePath -Value $releaseChannel -NoNewline
``` 
- Run the script => `releaseChannel.txt` under `$HOME\AppData\Roaming\JumpCloud Password Manager\data\daemon\releaseChannel.txt` should read `dogfood`

**For MacOs**
- Create the below `.sh` script on your machine
```shell
#!/bin/bash

# Set releaseChannel to beta OR dogfood OR public depending on your desired release channel
releaseChannel="dogfood"
#------- Do not modify below this line ------
allowed_values=("beta" "dogfood" "public")

if [[ ! " ${allowed_values[@]} " =~ " $releaseChannel " ]]; then
    echo "Error: Variable \$releaseChannel must be either 'beta', 'dogfood', or 'public'."
    exit 1
fi

for user in $(dscl . list /Users | grep -vE 'root|daemon|nobody|^_'); do
    if [[ -d /Users/$user ]]; then
        basePath="/Users/$user/Library/Application Support/JumpCloud Password Manager"
        filePath="$basePath/data/daemon/releaseChannel.txt"
        mkdir -p "$(dirname "$filePath")"
        echo -n "$releaseChannel" >"$filePath"
        sudo chown -R $user "$basePath"
    fi
done
``` 
- Run the script => `releaseChannel.txt` under `$HOME/Library/Application Support/JumpCloud Password Manager/data/daemon/releaseChannel.txt` should read `dogfood`

**For Linux**
- Create the below `.sh` script on your machine
```shell
#!/bin/bash

# Set $releaseChannel to beta OR dogfood OR public depending on your desired release channel
releaseChannel="dogfood"
#------- Do not modify below this line ------
allowed_values=("beta" "dogfood" "public")

if [[ ! " ${allowed_values[@]} " =~ " $releaseChannel " ]]; then
    echo "Error: Variable \$releaseChannel must be either 'beta', 'dogfood', or 'public'."
    exit 1
fi

for user in $(awk -F'[/:]' '{if ($3 >= 1000 && $3 != 65534) print $1}' /etc/passwd); do
    if [[ -d /home/$user ]]; then
        basePath="/home/$user/.config/JumpCloud Password Manager"
        filePath="$basePath/data/daemon/releaseChannel.txt"
        mkdir -p "$(dirname "$filePath")"
        echo -n "$releaseChannel" >"$filePath"
        sudo chown -R $user:$user "$basePath"
    fi
done
``` 
- Run the script => `releaseChannel.txt` under `$HOME/.config/JumpCloud Password Manager/data/daemon/releaseChannel.txt` should read `dogfood`

## Screenshots
N/A

[NB-11852]: https://jumpcloud.atlassian.net/browse/NB-11852?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

jworkmanjc commented:
@gbitarjc I'm not sure I'm following the point of this script in the public support repository.

Just testing the Windows and Mac scripts off the bat, neither seemed to work with production versions of the Password Manager installed. 

Windows seemed to fail outright even though there was an exit 0 status: 
![Screenshot 2024-06-24 at 1 30 41 PM](https://github.com/TheJumpCloud/support/assets/54448601/3fe8e2da-d1f1-4181-8063-dfc2753cbcc9)
MacOS gave me an exit 0 status code but there was no `$HOME/Library/Application Support/JC Password Manager Staging/data/daemon/releaseChannel.txt` file on my device

gbitarjc commented:
> @gbitarjc I'm not sure I'm following the point of this script in the public support repository.
> 
> Just testing the Windows and Mac scripts off the bat, neither seemed to work with production versions of the Password Manager installed.
> 
> Windows seemed to fail outright even though there was an exit 0 status: ![Screenshot 2024-06-24 at 1 30 41 PM](https://private-user-images.githubusercontent.com/54448601/342470325-3fe8e2da-d1f1-4181-8063-dfc2753cbcc9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk0ODI2MDUsIm5iZiI6MTcxOTQ4MjMwNSwicGF0aCI6Ii81NDQ0ODYwMS8zNDI0NzAzMjUtM2ZlOGUyZGEtZDFmMS00MTgxLTgwNjMtZGZjMjc1M2NiY2M5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjI3VDA5NTgyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVjY2M5YzdmMjRlYTQyZDQ2ZmZjYWM5ODYyMjc2ZDhlODc5MjliYWFlYWNmN2NkMzNlNTQzYjMwYTdmN2E0NTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.-oYqn3BvX-Q79rnQG1ATVqY-4Q4vMF7rqMSGMQHA2ss) MacOS gave me an exit 0 status code but there was no `$HOME/Library/Application Support/JC Password Manager Staging/data/daemon/releaseChannel.txt` file on my device



> @gbitarjc I'm not sure I'm following the point of this script in the public support repository.
> 
> Just testing the Windows and Mac scripts off the bat, neither seemed to work with production versions of the Password Manager installed.
> 
> Windows seemed to fail outright even though there was an exit 0 status: ![Screenshot 2024-06-24 at 1 30 41 PM](https://private-user-images.githubusercontent.com/54448601/342470325-3fe8e2da-d1f1-4181-8063-dfc2753cbcc9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk0ODI4MzUsIm5iZiI6MTcxOTQ4MjUzNSwicGF0aCI6Ii81NDQ0ODYwMS8zNDI0NzAzMjUtM2ZlOGUyZGEtZDFmMS00MTgxLTgwNjMtZGZjMjc1M2NiY2M5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjI3VDEwMDIxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc0MGVmZGYwOTU1Mzg2YzRhM2YzOWMwZDYzOWI2YzA4YjZlNzUzZWQ2ZmI2MDM4YjY3ZDU5ZTc1YWVlZWIyMTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.XWrIjMzWGaZLEPoD3QPxpCXqERR3eBYdQL3C3S-vEYM) MacOS gave me an exit 0 status code but there was no `$HOME/Library/Application Support/JC Password Manager Staging/data/daemon/releaseChannel.txt` file on my device

@jworkmanjc the purpose of these scripts is for admins to set the release channel for the password manager triggered through these scripts from the admin portal

also addressed the issues that you faced while testing and updated the scripts to only support production and updated the how to test section of this PR accordingly

jworkmanjc commented:
@gbitarjc 

There will be an issue with this command as it's currently designed. You've opened this PR to create command gallery scripts. 

The way most customers would use this would be to create the command in the JumpCloud console and deploy it to devices. Most customers are going to assume that they can run these commands and that it will do the intended action described. 

At least for password manager there's an additional complexity. The PWM tool is installed per user. These scripts that set the release channel should explain this. If you run this script through a JumpCloud command it's not going to do anything because those commands run as root. 

Is there some need from individual users to set the release channel? Personally I can't imagine an individual user (not customer, user of password manager) setting their release channel for any reason. The only thing I could see is that some JumpCloud administrator wanting to set the release channel for their end users. If that's the case, and the intention of this PR, these scripts need to be re-written to change the release channel for every user that's installed PWM on a device.

If this script is intended to be run by individual users, then it shouldn't be a command gallery script in my opinion. At the moment I think this is going to confuse our administrators. 

Running this script on macOS returns exit code 0 but does not update the release channel for the any user on the system. 
<img width="976" alt="Screenshot 2024-06-27 at 11 19 50 AM" src="https://github.com/TheJumpCloud/support/assets/54448601/26b54743-c270-487f-97dc-ac52c2729080">

gbitarjc commented:
@jworkmanjc 

After taking into consideration that the PWM tool is installed per user. the scripts were adjusted to cater for every user.

There is no need from individual users to set the release channel, these scripts are for  JumpCloud administrators that are wanting to set the release channel for their end users.

jworkmanjc commented on PowerShell/JumpCloud Commands Gallery/Windows Commands/Windows - Set ReleaseChannel In JumpCloud Password Manager.md:
````diff
@@ -0,0 +1,35 @@
+
+```
+# Set $RELEASE_CHANNEL to beta OR dogfood OR public depending on your desired release channel
+$RELEASE_CHANNEL = "public"
````
We can't expect admins to do the right thing, it would probably be a good idea to add a check to ensure that `$RELEASE_CHANNEL` is one of 'public','dogfood','beta'

jworkmanjc commented on PowerShell/JumpCloud Commands Gallery/Mac Commands/Mac - Set ReleaseChannel In JumpCloud Password Manager.md:
```diff
@@ -0,0 +1,44 @@
+#!/bin/bash
+
+# Set $RELEASE_CHANNEL to beta OR dogfood OR public ON LINE 15 depending on your desired release channel
+RELEASE_CHANNEL="public"
```
We can't expect admins to do the right thing, it would probably be a good idea to add a check to ensure that `$RELEASE_CHANNEL` is one of 'public','dogfood','beta'

jworkmanjc commented on PowerShell/JumpCloud Commands Gallery/Linux Commands/Linux - Set ReleaseChannel In JumpCloud Password Manager.md:
```diff
@@ -0,0 +1,45 @@
+
+# Set $RELEASE_CHANNEL to beta OR dogfood OR public ON LINE 4 depending on your desired release channel
+
+RELEASE_CHANNEL="public"
```
We can't expect admins to do the right thing, it would probably be a good idea to add a check to ensure that `$RELEASE_CHANNEL` is one of 'public','dogfood','beta'

jworkmanjc commented on PowerShell/JumpCloud Commands Gallery/Mac Commands/Mac - Set ReleaseChannel In JumpCloud Password Manager.md:
````diff
@@ -0,0 +1,44 @@
+```
+#!/bin/bash
+
+# Set $RELEASE_CHANNEL to beta OR dogfood OR public ON LINE 15 depending on your desired release channel
````
LINE 15?

jworkmanjc commented on PowerShell/JumpCloud Commands Gallery/Linux Commands/Linux - Set ReleaseChannel In JumpCloud Password Manager.md:
````diff
@@ -0,0 +1,45 @@
+```
+#!/bin/bash
+
+# Set $RELEASE_CHANNEL to beta OR dogfood OR public ON LINE 4 depending on your desired release channel
````
LINE 4? I think we should just say change the variable value. If you copy and paste this today, it's LINE 5

jworkmanjc commented on PowerShell/JumpCloud Commands Gallery/Windows Commands/Windows - Set ReleaseChannel In JumpCloud Password Manager.md:
````diff
@@ -0,0 +1,35 @@
+#### Command
+
+```
+# Set $RELEASE_CHANNEL to beta OR dogfood OR public depending on your desired release channel
````
We are not noting the line here, and I think that's the correct way to tell customers what to do, since it's really easy to forget that we need to update that in the future.

jworkmanjc commented on PowerShell/JumpCloud Commands Gallery/Windows Commands/Windows - Set ReleaseChannel In JumpCloud Password Manager.md:
```diff
@@ -0,0 +1,35 @@
+
+#------- Do not modify below this line ------
+
+$FILE_PATH = "$env:APPDATA\JumpCloud Password Manager\data\daemon\releaseChannel.txt"
```
I'd encourage you to go run a Windows PowerShell command from JumpCloud and print out the `$ENV:APPDATA` directory. 

The windows command doesn't work as described in the PR.
![Screenshot 2024-07-09 at 8 20 02 AM](https://github.com/TheJumpCloud/support/assets/54448601/d6e1e23f-051a-4e85-b2d3-d667da7b7e4a)

Look at this command and I think you'll see why.
![Screenshot 2024-07-09 at 8 21 23 AM](https://github.com/TheJumpCloud/support/assets/54448601/46796d8c-0e3a-4bd2-ab88-629f14b814b5)

jworkmanjc reviewed (changes requested):
I haven't tested the Linux and MacOS commands yet. Windows I know is not working. I'll take additional time to validate the Linux and MacOS commands when the Windows command is re-worked to function as described in the Pull Request.

jworkmanjc reviewed (dismissed):
![Screenshot 2024-07-10 at 10 53 00 AM](https://github.com/TheJumpCloud/support/assets/54448601/a2d295ca-426b-4ebf-a27c-d8eebacf9128)
![Screenshot 2024-07-10 at 10 46 40 AM](https://github.com/TheJumpCloud/support/assets/54448601/8a3ce8a2-812c-4bdf-bec2-78288326da4d)
![Screenshot 2024-07-10 at 10 41 49 AM](https://github.com/TheJumpCloud/support/assets/54448601/a20baf24-42d9-4a1b-86f4-26a554589e8d)

Validated windows, mac, linux. 

My only concern is that there will be some customer who states that the linux/mac commands work differently than the Windows command which relies on the user being logged in to set the release channel. Overall I doubt it'll come up often if at all but I'm calling it out as a potential issue that would need to be resolved later.

Otherwise these can all be invoked through JumpCloud commands. Thank you @gbitarjc for the many iterations I've requested of you. It'll go a long ways towards preventing support calls and other customer issues.

jworkmanjc commented on PowerShell/JumpCloud Commands Gallery/Mac Commands/Mac - Set ReleaseChannel In JumpCloud Password Manager.md:
```diff
@@ -0,0 +1,51 @@
+#!/bin/bash
+
+# Set releaseChannel to beta OR dogfood OR public depending on your desired release channel
+releaseChannel="dogfood"
```
Let's change this to public so that every script is set to the public channel by default? 
```suggestion
releaseChannel="public"
```

gweinjc reviewed (dismissed):
Tested on Windows, Linux and Mac. Appears to be working on my devices

gbitarjc commented:
@jworkmanjc, thank you for your help with this PR i really appreciated the help.

jworkmanjc reviewed (approved):
Looks good thank you for all the changes :)
